packages/nextjs/hooks/scaffold-eth/useDeployedContractInfo.ts

```diff
@@ -1,52 +1,52 @@
-import { useEffect, useState } from "react";
+import { useEffect, useMemo, useState } from "react";
 import { useProvider } from "wagmi";
 import { getTargetNetwork } from "~~/utils/scaffold-eth";
-type GeneratedContractType = {
-  address: string;
-  abi: any[];
-};
+import contracts from "../../generated/hardhat_contracts";
+import { Contract, ContractName, ContractCodeStatus, DefaultChain } from "./contract.types";
 
 /**
  * @dev use this hook to get a deployed contract from `yarn deploy` generated files.
  * @param contractName - name of deployed contract
  * @returns {GeneratedContractType | undefined} object containing contract address and abi or undefined if contract is not found
  */
-export const useDeployedContractInfo = (contractName: string | undefined | null) => {
+export const useDeployedContractInfo = <TContractName extends ContractName>(contractName: TContractName) => {
   const configuredChain = getTargetNetwork();
-  const [deployedContractData, setDeployedContractData] = useState<undefined | GeneratedContractType>(undefined);
-  const [isLoading, setIsLoading] = useState(true);
+  const deployedContract = useMemo(
+    () =>
+      contracts[`${configuredChain.id}` as DefaultChain]?.[0]?.contracts?.[
+        contractName as ContractName
+      ] as Contract<TContractName>,
+    [configuredChain.id, contractName],
+  );
+  const [status, setStatus] = useState<ContractCodeStatus>(ContractCodeStatus.LOADING);
   const provider = useProvider({ chainId: configuredChain.id });
 
   useEffect(() => {
-    const getDeployedContractInfo = async () => {
-      setIsLoading(true);
-      let ContractData;
-      try {
-        ContractData = require("~~/generated/hardhat_contracts.json");
-        const contractsAtChain = ContractData[configuredChain.id as keyof typeof ContractData];
-        const contractsData = contractsAtChain?.[0]?.contracts;
-        const deployedContract = contractsData?.[contractName as keyof typeof contractsData];
+    let mounted = true;
 
-        if (!deployedContract || !contractName || !provider) {
-          return;
-        }
+    const checkContractDeployment = async () => {
+      const code = await provider.getCode((deployedContract as Contract<TContractName>).address);
 
-        const code = await provider.getCode(deployedContract.address);
-        // If contract code is `0x` => no contract deployed on that address
-        if (code === "0x" || !contractsData || !(contractName in contractsData)) {
-          return;
-        }
-        setDeployedContractData(contractsData[contractName]);
-      } catch (e) {
-        // Contract not deployed or file doesn't exist.
-        setDeployedContractData(undefined);
-      } finally {
-        setIsLoading(false);
+      if (!mounted) {
```

Comment:
not sure how mounted can be `false` here

Reply:
If the hook is used in a component that gets mounted and unmounted before the async call resolves (e.g. user navigates to another page).

The pattern with the `mounted` is often used when a `useEffect` needs to call async functionality to avoid doing state changes on unmounted components.

In short, it's not always technically necessary, but it's good practice, and it ensures the console doesn't get random warnings like the following:

> Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.

Reply:
I understood you did that because of that warning, but I didn't meet that pattern before. At least I don't remember. Nice one, thanks :)

Reply:
Didn't know about the pattern either! I'll try not to overuse it now haha

https://stackoverflow.com/a/60907638

Reply:
PS : we have great util hook alerady for this instead of implementing on our own -> https://usehooks-ts.com/react-hook/use-is-mounted

Reply:
Ah cool, for consistency, I guess it's cleaner to use the hook. Reduces boilerplate :)

I'll update it